Python/bytecodes.c

```diff
@@ -2114,11 +2115,34 @@ dummy_func(
         }
 
         inst(JUMP_BACKWARD, (--)) {
-            assert(oparg < INSTR_OFFSET());
-            JUMPBY(-oparg);
+            _Py_CODEUNIT *here = next_instr-1;
+            assert(oparg <= INSTR_OFFSET());
+            JUMPBY(1-oparg);
+            here[1].cache += (1 << OPTIMIZER_BITS_IN_COUNTER);
+            if (here[1].cache > tstate->interp->optimizer_backedge_threshold) {
+                OBJECT_STAT_INC(optimization_attempts);
+                frame = _PyOptimizer_BackEdge(frame, here, next_instr, stack_pointer);
+                if (frame == NULL) {
+                    frame = cframe.current_frame;
+                    goto error;
+                }
+                here[1].cache &= ((1 << OPTIMIZER_BITS_IN_COUNTER) -1);
```

Comment:
> What is the purpose of OPTIMIZER_BITS_IN_COUNTER?

To provide some bits for the optimizer to use. We don't need all 16 bits for a counter, so the optimizer can use them for per-hotspot data. I plan to use them for rate-limiting the optimizer.

Reply:
I'm a bit confused about how an optimizer would actually make use of these four bits. The optimizer only gets invoked when its threshold is met, and when the optimizer is invoked, it is required to return an executor (returning `NULL` signals an exception). The executor will generally be per-hotspot, and can have as much custom state in it as the optimizer wants.

I guess the use case might be that sometimes an optimizer will choose to set a shared no-op executor for some reason (so if it's shared, it can't store per-hotspot state in it), but still want to store some per-hotspot state?

Reply:
One possibility is to use them as a counter to compare against a global counter to avoid optimizing code that reaches the threshold, but slowly, so isn't "hot". I plan to do something like this for our optimizer.
Another is to serve as a marker to prevent optimization where is has proved to not be beneficial.

Executors cannot be shared as they hold the original opcode and oparg.

Reply:
In order to better support the use cases you mention, shouldn't optimizers have the option to not return an executor from `optimize` (without signalling an error)?

Currently it seems like `optimize` must return an executor, and as you point out executors can't be shared, in which case the optimizer can put whatever state it wants into the executor and doesn't need these inline bits?

Reply:
Yes, we should check for `NULL`.